### PR TITLE
Optimized compilation for arrays when using a['b'] notation

### PR DESCRIPTION
**Pull request by fabpot**
Under certain conditions, and thanks to the `??` operator, it's possible to optimize array calls (and avoid the call to `twig_get_attribute()`).

On my local machine, code is almost twice as fast, which is nice if you are using a lot of `a["b"]` calls of nested ones.

refs #2547

### CODE REVIEW

File: lib/Twig/Node/Expression/GetAttr.php [hunk 1]
```diff
@@ -23,6 +23,25 @@ public function __construct(Twig_Node_Expression $node, Twig_Node_Expression $at
...
     public function compile(Twig_Compiler $compiler)
     {
+        // optimize array calls
+        if (
+            (!$compiler->getEnvironment()->isStrictVariables() || $this->getAttribute('ignore_strict_check'))
+            && !$this->getAttribute('is_defined_test')
+            && Twig_Template::ARRAY_CALL === $this->getAttribute('type')
+        ) {
+            $compiler
+                ->raw('(is_array(')
```
Comment (stof) [inline lib/Twig/Node/Expression/GetAttr.php]: shouldn't it be `is_array($node) || $node instanceof ArrayAccess` ?

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
btw, the fact that Travis is green makes me wonder whether we have enough integration tests for the features implemented by GetAttrNode

File: lib/Twig/Node/Expression/GetAttr.php (hunk 1 above)
Comment (fabpot) [reply lib/Twig/Node/Expression/GetAttr.php]: indeed, that's even better

Comment (jvasseur) [inline lib/Twig/Node/Expression/GetAttr.php]: Isn't this missing handling of objects implementing `ArrayAccess` ?

**Comment by fabpot** [comment]
@stof I don't understand your comment about tests. We have plenty of tests on this part of Twig. A lot of edge cases are tested as well and we added a lot of them when we worked on the C ext.

**Comment by stof** [comment]
@fabpot but your initial version of the PR was breaking support for ARRAY_CALL on an ArrayAccess object when strict checks are not enabled, and nothing was failing in the CI. So this is not covered by tests.

**Comment by fabpot** [comment]
Indeed, there were no tests for compiled code when strict_variables was false. Tests added.

File: test/Twig/Tests/Node/Expression/GetAttrTest.php [hunk 2]
```diff
@@ -37,7 +37,7 @@ public function getTests()
         $tests[] = array($node, sprintf('%s%s, "bar", array())', $this->getAttributeGetter(), $this->getVariableGetter('foo', 1)));
 
         $node = new Twig_Node_Expression_GetAttr($expr, $attr, $args, Twig_Template::ARRAY_CALL, 1);
-        $tests[] = array($node, sprintf('%s%s, "bar", array(), "array")', $this->getAttributeGetter(), $this->getVariableGetter('foo', 1)));
+        $tests[] = array($node, sprintf('(is_array(%s) || ($context["foo"] ?? null) instanceof ArrayAccess ? (($context["foo"] ?? null)["bar"] ?? null) : null)', $this->getVariableGetter('foo', 1), $this->getVariableGetter('foo', 1)));
```
Comment (stof) [inline test/Twig/Tests/Node/Expression/GetAttrTest.php]: shouldn't `($context["foo"] ?? null)` use `$this->getVariableGetter` here too ?

**Review by stof — CHANGES_REQUESTED** [review]

File: test/Twig/Tests/Node/Expression/GetAttrTest.php (hunk 2 above)
Comment (fabpot) [reply test/Twig/Tests/Node/Expression/GetAttrTest.php]: These abstractions were useful in 1.x, not so much now. I've kept them to ease merging, but no need to keep them for new code.

**Comment by fabpot** [comment]
My first optimization for simple calls made performance worse in case of a deeply nested array.

I've just optimized calls to those deeply nested array calls (like `a["b"]["c"]["d"]`).  Now both are much faster than before.

I've updated the `twig/twig-perf` repo to add a new test for nested array calls so that we can track regressions.

Here are the results:

```
                    |      1.x |      2.x | array-perf-optim |
simple_array_access |    204.7 |    361.7 |             51.1 |
nested_array_access |    491.3 |    782.9 |             80.8 |
```
